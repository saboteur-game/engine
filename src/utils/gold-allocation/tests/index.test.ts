import RewardDeck from "../../../models/reward-deck";
import Player from "../../../models/player";
import {
  IPlayers,
  allocateGoldToGoldDiggers,
  allocateGoldToSaboteurs,
} from "../";

describe("Gold allocation", () => {
  let rewardDeck: RewardDeck;
  let playOrder: string[];
  let players: IPlayers;
  let lastPlayer: Player;

  const generatePlayers = (count: number, saboteurs: number): void => {
    const gamePlayers = Array(count)
      .fill(undefined)
      .map((value, index) => new Player(`Player ${index + 1}`, 20 + index));

    for (let i = 1; i <= saboteurs; i++) {
      gamePlayers[i * 2].setup(true);
    }

    players = gamePlayers.reduce(
      (mapping, player) => ({
        ...mapping,
        [player.id]: player,
      }),
      {}
    );
    playOrder = gamePlayers.map(({ id }) => id);
  };

  beforeEach(() => {
    rewardDeck = new RewardDeck();
  });

  describe("allocateGoldToGoldDiggers", () => {
    describe("when the saboteur is the active player", () => {
      beforeEach(() => {
        generatePlayers(6, 2);
        lastPlayer = players[playOrder[2]];
      });

      it("allocates gold to the gold diggers", () => {
        const allocatedGold = allocateGoldToGoldDiggers(
          rewardDeck,
          playOrder,
          players,
          lastPlayer
        );
        expect(allocatedGold[lastPlayer.id]).toBe(undefined);
        expect(allocatedGold).toMatchSnapshot();
      });
    });

    describe("when the reward deck is empty", () => {
      beforeEach(() => {
        const totalCards = rewardDeck.getCardCount();
        for (let i = 0; i < totalCards; i++) rewardDeck.drawCard();
        generatePlayers(4, 0);
        lastPlayer = players[playOrder[0]];
      });

      it("does not allocate any gold", () => {
        expect(
          allocateGoldToGoldDiggers(rewardDeck, playOrder, players, lastPlayer)
        ).toStrictEqual({});
      });
    });

    describe("when there are no saboteurs", () => {
      beforeEach(() => {
        generatePlayers(4, 0);
        lastPlayer = players[playOrder[0]];
      });

      it("allocates gold to the gold diggers", () => {
        expect(
          allocateGoldToGoldDiggers(rewardDeck, playOrder, players, lastPlayer)
        ).toMatchSnapshot();
      });
    });

    describe("when there is 1 saboteur", () => {
      beforeEach(() => {
        generatePlayers(5, 1);
      });

      it("allocates gold to the gold diggers", () => {
        expect(
          allocateGoldToGoldDiggers(rewardDeck, playOrder, players, lastPlayer)
        ).toMatchSnapshot();
      });
    });

    describe("when there are 2 saboteurs", () => {
      beforeEach(() => {
        generatePlayers(6, 2);
      });

      it("allocates gold to the gold diggers", () => {
        expect(
          allocateGoldToGoldDiggers(rewardDeck, playOrder, players, lastPlayer)
        ).toMatchSnapshot();
      });
    });

    describe("when there are 3 saboteurs", () => {
      beforeEach(() => {
        generatePlayers(7, 3);
      });

      it("allocates gold to the gold diggers", () => {
        expect(
          allocateGoldToGoldDiggers(rewardDeck, playOrder, players, lastPlayer)
        ).toMatchSnapshot();
      });
    });

    describe("when there are 4 saboteurs", () => {
      beforeEach(() => {
        generatePlayers(10, 4);
      });

      it("allocates gold to the gold diggers", () => {
        expect(
          allocateGoldToGoldDiggers(rewardDeck, playOrder, players, lastPlayer)
        ).toMatchSnapshot();
      });
    });
  });

  describe("allocateGoldToSaboteurs", () => {
    describe("when there are no saboteurs", () => {
      beforeEach(() => {
        generatePlayers(4, 0);
      });

      it("allocates no gold", () => {
        expect(
          allocateGoldToSaboteurs(rewardDeck, playOrder, players)
        ).toStrictEqual({});
      });
    });

    describe("when there is 1 saboteur", () => {
      beforeEach(() => {
        generatePlayers(5, 1);
      });

      it("allocates gold to the saboteurs", () => {
        expect(
          allocateGoldToSaboteurs(rewardDeck, playOrder, players)
        ).toMatchSnapshot();
      });
    });

    describe("when there are 2 saboteurs", () => {
      beforeEach(() => {
        generatePlayers(6, 2);
      });

      it("allocates gold to the saboteurs", () => {
        expect(
          allocateGoldToSaboteurs(rewardDeck, playOrder, players)
        ).toMatchSnapshot();
      });
    });

    describe("when there are 3 saboteurs", () => {
      beforeEach(() => {
        generatePlayers(7, 3);
      });

      it("allocates gold to the saboteurs", () => {
        expect(
          allocateGoldToSaboteurs(rewardDeck, playOrder, players)
        ).toMatchSnapshot();
      });
    });

    describe("when there are 4 saboteurs", () => {
      beforeEach(() => {
        generatePlayers(10, 4);
      });

      it("allocates gold to the saboteurs", () => {
        expect(
          allocateGoldToSaboteurs(rewardDeck, playOrder, players)
        ).toMatchSnapshot();
      });
    });
  });
});
