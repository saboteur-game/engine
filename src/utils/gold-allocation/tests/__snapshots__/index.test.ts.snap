// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gold allocation allocateGoldToGoldDiggers when the saboteur is the active player allocates gold to the gold diggers 1`] = `
Object {
  "test-id-29": Array [
    Object {
      "id": "test-id-2",
      "value": 3,
    },
  ],
  "test-id-30": Array [
    Object {
      "id": "test-id-1",
      "value": 3,
    },
    Object {
      "id": "test-id-4",
      "value": 3,
    },
  ],
  "test-id-32": Array [
    Object {
      "id": "test-id-3",
      "value": 3,
    },
    Object {
      "id": "test-id-6",
      "value": 2,
    },
  ],
  "test-id-34": Array [
    Object {
      "id": "test-id-5",
      "value": 2,
    },
  ],
}
`;

exports[`Gold allocation allocateGoldToGoldDiggers when there are 2 saboteurs allocates gold to the gold diggers 1`] = `
Object {
  "test-id-29": Array [
    Object {
      "id": "test-id-1",
      "value": 3,
    },
  ],
  "test-id-30": Array [
    Object {
      "id": "test-id-3",
      "value": 3,
    },
    Object {
      "id": "test-id-6",
      "value": 2,
    },
  ],
  "test-id-32": Array [
    Object {
      "id": "test-id-2",
      "value": 3,
    },
    Object {
      "id": "test-id-5",
      "value": 2,
    },
  ],
  "test-id-34": Array [
    Object {
      "id": "test-id-4",
      "value": 3,
    },
  ],
}
`;

exports[`Gold allocation allocateGoldToGoldDiggers when there are 3 saboteurs allocates gold to the gold diggers 1`] = `
Object {
  "test-id-29": Array [
    Object {
      "id": "test-id-1",
      "value": 3,
    },
  ],
  "test-id-30": Array [
    Object {
      "id": "test-id-3",
      "value": 3,
    },
    Object {
      "id": "test-id-6",
      "value": 2,
    },
  ],
  "test-id-32": Array [
    Object {
      "id": "test-id-2",
      "value": 3,
    },
    Object {
      "id": "test-id-5",
      "value": 2,
    },
  ],
  "test-id-34": Array [
    Object {
      "id": "test-id-4",
      "value": 3,
    },
    Object {
      "id": "test-id-7",
      "value": 2,
    },
  ],
}
`;

exports[`Gold allocation allocateGoldToGoldDiggers when there are 4 saboteurs allocates gold to the gold diggers 1`] = `
Object {
  "test-id-29": Array [
    Object {
      "id": "test-id-1",
      "value": 3,
    },
  ],
  "test-id-30": Array [
    Object {
      "id": "test-id-5",
      "value": 2,
    },
  ],
  "test-id-32": Array [
    Object {
      "id": "test-id-4",
      "value": 3,
    },
    Object {
      "id": "test-id-9",
      "value": 2,
    },
  ],
  "test-id-34": Array [
    Object {
      "id": "test-id-3",
      "value": 3,
    },
    Object {
      "id": "test-id-8",
      "value": 2,
    },
  ],
  "test-id-36": Array [
    Object {
      "id": "test-id-2",
      "value": 3,
    },
    Object {
      "id": "test-id-7",
      "value": 2,
    },
  ],
  "test-id-38": Array [
    Object {
      "id": "test-id-6",
      "value": 2,
    },
  ],
}
`;

exports[`Gold allocation allocateGoldToGoldDiggers when there are no saboteurs allocates gold to the gold diggers 1`] = `
Object {
  "test-id-29": Array [
    Object {
      "id": "test-id-1",
      "value": 3,
    },
  ],
  "test-id-30": Array [
    Object {
      "id": "test-id-3",
      "value": 3,
    },
  ],
  "test-id-31": Array [
    Object {
      "id": "test-id-2",
      "value": 3,
    },
  ],
  "test-id-32": Array [
    Object {
      "id": "test-id-4",
      "value": 3,
    },
  ],
}
`;

exports[`Gold allocation allocateGoldToGoldDiggers when there is 1 saboteur allocates gold to the gold diggers 1`] = `
Object {
  "test-id-29": Array [
    Object {
      "id": "test-id-1",
      "value": 3,
    },
  ],
  "test-id-30": Array [
    Object {
      "id": "test-id-3",
      "value": 3,
    },
  ],
  "test-id-32": Array [
    Object {
      "id": "test-id-2",
      "value": 3,
    },
    Object {
      "id": "test-id-5",
      "value": 2,
    },
  ],
  "test-id-33": Array [
    Object {
      "id": "test-id-4",
      "value": 3,
    },
  ],
}
`;

exports[`Gold allocation allocateGoldToSaboteurs when there are 2 saboteurs allocates gold to the saboteurs 1`] = `
Object {
  "test-id-31": Array [
    Object {
      "id": "test-id-1",
      "value": 3,
    },
  ],
  "test-id-33": Array [
    Object {
      "id": "test-id-2",
      "value": 3,
    },
  ],
}
`;

exports[`Gold allocation allocateGoldToSaboteurs when there are 3 saboteurs allocates gold to the saboteurs 1`] = `
Object {
  "test-id-31": Array [
    Object {
      "id": "test-id-1",
      "value": 3,
    },
  ],
  "test-id-33": Array [
    Object {
      "id": "test-id-2",
      "value": 3,
    },
  ],
  "test-id-35": Array [
    Object {
      "id": "test-id-3",
      "value": 3,
    },
  ],
}
`;

exports[`Gold allocation allocateGoldToSaboteurs when there are 4 saboteurs allocates gold to the saboteurs 1`] = `
Object {
  "test-id-31": Array [
    Object {
      "id": "test-id-5",
      "value": 2,
    },
  ],
  "test-id-33": Array [
    Object {
      "id": "test-id-6",
      "value": 2,
    },
  ],
  "test-id-35": Array [
    Object {
      "id": "test-id-7",
      "value": 2,
    },
  ],
  "test-id-37": Array [
    Object {
      "id": "test-id-8",
      "value": 2,
    },
  ],
}
`;

exports[`Gold allocation allocateGoldToSaboteurs when there is 1 saboteur allocates gold to the saboteurs 1`] = `
Object {
  "test-id-31": Array [
    Object {
      "id": "test-id-1",
      "value": 3,
    },
    Object {
      "id": "test-id-13",
      "value": 1,
    },
  ],
}
`;
